src/main/java/org/assertj/core/description/JoinDescription.java

```diff
@@ -0,0 +1,214 @@
+  private static Description nonNull(Description description) {
+    Preconditions.checkNotNull(description, "The descriptions should not contain null elements");
+    return description;
+  }
+
+  @Override
+  public String value() {
+    return value(new IndentedAppendable(new StringBuilder(), INDENTION)).toString();
+  }
+
+  @Override
+  public int hashCode() {
+    return HASH_CODE_PRIME + hashCodeFor(prefix) + hashCodeFor(suffix) + hashCodeFor(descriptions);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (!(o instanceof JoinDescription)) return false;
+    JoinDescription that = (JoinDescription) o;
+    return areEqual(descriptions, that.descriptions) &&
+      areEqual(prefix, that.prefix) &&
+      areEqual(suffix, that.suffix);
+  }
+
+  private IndentedAppendable value(IndentedAppendable buff) {
+    // shallow indention for prefix
+    buff.indentDelta(-INDENTION)
+      .indent()
+      .append(prefix);
+
+    if (descriptions.isEmpty()) {
+      return buff.append(suffix); // no line sep
+    }
+
+    buff.append(LINE_SEP).indentDelta(INDENTION);
+
+    Iterator<? extends Description> it = descriptions.iterator();
+
+    while (it.hasNext()) {
+      Description desc = it.next();
+
+      if (desc instanceof JoinDescription) {
+        JoinDescription joinDesc = (JoinDescription) desc;
+
+        // increase indention to write nested conditions more deeply
+        joinDesc.value(buff.indentDelta(INDENTION));
+      } else {
+        buff.indent().append(desc.value());
+      }
+
+      // avoid the trailing delimiter
+      if (it.hasNext()) {
+        buff.append(DELIMITER);
+      }
+    }
+
+    return buff.append(LINE_SEP)
+      .indentDelta(-INDENTION) // shallow indention to align with prefix
+      .indent()
+      .append(suffix);
+  }
+
+  /**
+   * The wrapper for {@code StringBuilder} aware of indention.
+   */
+  private static class IndentedAppendable implements Appendable {
+    private static final String[] CACHE = {
+      "",
+      " ",
+      "  ",
+      "   ",
+      "    ",
+      "     ",
+      "      ",
+      "       ",
+      "        ",
+      "         ",
+    };
+    private final StringBuilder buff;
+    private int size;
+
+    public IndentedAppendable(StringBuilder buff, int size) {
+      this.buff = buff;
+      this.size = size;
+    }
+
+    @Override
+    public IndentedAppendable append(CharSequence csq) {
+      buff.append(csq);
+      return this;
+    }
+
+    @Override
+    public IndentedAppendable append(CharSequence csq, int start, int end) {
+      buff.append(csq, start, end);
+      return this;
+    }
+
+    @Override
+    public IndentedAppendable append(char c) {
+      buff.append(c);
+      return this;
+    }
+
+    /**
+     * Adjusts the indention size by {@code delta}.
+     *
+     * @param delta The indention adjustment.
+     *
+     * @return a this instance.
+     */
+    IndentedAppendable indentDelta(int delta) {
```

Comment:
Don't you find it bit misleading? I mean the word `increase` because we decreasing it also. Can't find anything better then `delta`.

Reply:
fair enough, `changeIndentationBy` should do

Reply:
question: is there a reason we don't indent straight away in this method ?
would it make sense to replace this method and `indent()` by `indentBy(int indentation)` which would directly apply the given indentation (negative or positive).

The code would be a bit easier to read, eg :
```java
 buff.append(LINE_SEPARATOR)
     .indentDelta(-DEFAULT_INDENTATION) // shallow indention to align with prefix
     .indent()
     .append(suffix);
```
would be changed to:
```java
 buff.append(LINE_SEPARATOR)
     .indentBy(-DEFAULT_INDENTATION) // shallow indention to align with prefix
     .append(suffix);
```

Reply:
> would it make sense to replace this method and indent() by indentBy(int indentation)

Exactly, initially I was thinking in a same way, which make sens of course. I have implemented but PR was not ready yet. I come back to proceed with PR, it turns out code incredibly hard to understand (with `indentBy(int indentation)`). It requires quite big amount of magic and tricky pre-assumptions, because of *recursive* nature `value(IndentedAppendable)` method.

Reply:
Although I was trying implement it with single method responsible for the size and actual indentation.

Reply:
alright, let's keep the code as simple as possible and use `changeIndentationBy` along with `indent` then